keys/employees-keys/README.md: major redesign of the procedure (#101)
3mdeb/3mdeb-secpack: keys/employees-keys/README.md

This directory keeps 3mdeb employees keys and explains the procedure for adding
new keys.

## Table of contents

* [Adding key to the repository](#adding-key-to-the-repository)
* [FAQ](#faq)
  - [What to do when this key expires?](#what-to-do-when-this-key-expires)
  - [If we extend expiry every time, why not set "key does not
    expire"?](#if-we-extend-expiry-every-time-why-not-set-key-does-not-expire)

>   - [What to do when private key was lost or

Reviewer: This link does not render correctly

Author: I have no idea why this doesn't work and previous works fine: cca3480e6135

Reviewer 2: @pietrushnic @macpijan there is no such section as `What to do when private key was lost or compromised?` that is why link does not work.

Author: Ok, fixed 282ec0ab04ea

Author: @macpijan can we close this conversation?

Author: @macpijan ping?

## Adding key to the repository

To add new key to this repository you should:

1. [Create new GPG key](#how-to-create-a-new-gpg-key)
2. [Encrypt revocation cert](#encrypt-revocation-cert)
3. [Send key for signing and revocation
   cert](#sending-key-for-signing-and-revocation-cert)

Then 3mdeb Team Leader or Management should

1. [3mdeb Team Leader/Management: Identity
   confirmation](#identity-confirmation)
2. [3mdeb Team Leader/Management: Key signing](#key-signing)

Finally you should configure your git and provide signed key through pull
request:

1. [Git configuration](#git-configuration)
2. [Create Pull Request with signed key](#create-a-pull-request-with-key)
3. [Upload to keys.opengpg.org](#upload-to-keys-openpgp-org)

If you have questions, please check [FAQ](#FAQ) section. Maybe the answer is
already there. If not, feel free to submit an issue to the repository.

### Create a new GPG key

The following assumes that you have `gpg` installed.  The interaction is broken
into sections that correspond to user queries.

Clone repository:

```shell
git clone https://github.com/3mdeb/3mdeb-secpack.git
```

Change directory:

```shell

> cd 3mdeb-secpack/scripts

Reviewer: Do we have to execute it from `scripts`, not the root dir?

In a typical flow, we first use this script, and then move to the next section. We would need to go back one directory level so commands from the [Encrypt revocation cert](https://github.com/3mdeb/3mdeb-secpack/blob/5b47704c3380fd7f773dc425114ba4ae9f7cba4c/keys/employees-keys/README.md#encrypt-revocation-cert) work fine

Author: @macpijan, from testing the most recent version, it doesn't matter which location we run the script. The directory should be writable by user-run script since we create temporary files. Do you expect me to change anything here?

Author: @macpijan ping?

```

Generate key:

```shell

> ./gen-new-employee-key.sh "Your Name" "your.name@3mdeb.com"

Reviewer: For me, it ends with an Error. Here is a debug output from script:

```
./gen-new-employee-key.sh "Your Name" "your.name@3mdeb.com"
+ '[' 2 -ne 2 ']'
+ KEY_NAME='Your Name'
+ KEY_EMAIL=your.name@3mdeb.com
+ KEY_COMMENT='Employee Cert Key'
+ KEY_FILE=your.name@3mdeb.com.asc
+ REVOCATION_FILE=your.name@3mdeb.com-revocation.asc
++ cat
+ BATCH_FILE_CONTENT='%echo Generating a primary certification key and subkeys for signing and encryption
Key-Type: RSA
Key-Length: 4096
Key-Usage: cert
Expire-Date: 0
Subkey-Type: RSA
Subkey-Length: 4096
Subkey-Usage: sign
Name-Real: Your Name
Name-Comment: Employee Cert Key
Name-Email: your.name@3mdeb.com
%no-protection
%commit
%echo Key generation complete'
+ echo '%echo Generating a primary certification key and subkeys for signing and encryption
Key-Type: RSA
Key-Length: 4096
Key-Usage: cert
Expire-Date: 0
Subkey-Type: RSA
Subkey-Length: 4096
Subkey-Usage: sign
Name-Real: Your Name
Name-Comment: Employee Cert Key
Name-Email: your.name@3mdeb.com
%no-protection
%commit
%echo Key generation complete'
++ gpg --batch --generate-key temp_batch_file.txt
++ grep 'gpg: key'
++ awk '{print $3}'
+ KEY_ID=
+ '[' -z '' ']'
+ echo 'Error: Could not retrieve KEY_ID.'
Error: Could not retrieve KEY_ID.
+ exit 1
```

Reviewer: ```
gpg --version
gpg (GnuPG) 2.4.0
libgcrypt 1.10.2-unknown
Copyright (C) 2021 Free Software Foundation, Inc.
License GNU GPL-3.0-or-later <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Home: /home/macpijan/.gnupg
Supported algorithms:
Pubkey: RSA, ELG, DSA, ECDH, ECDSA, EDDSA
Cipher: IDEA, 3DES, CAST5, BLOWFISH, AES, AES192, AES256, TWOFISH,
        CAMELLIA128, CAMELLIA192, CAMELLIA256
Hash: SHA1, RIPEMD160, SHA256, SHA384, SHA512, SHA224
Compression: Uncompressed, ZIP, ZLIB, BZIP2
```

Reviewer: @pietrushnic  Can you show the debug output on your side, so we can more easily find what the problem is?

Author: @macpijan I assume this is GPG version issue:

```
gpg (GnuPG) 2.2.40
libgcrypt 1.10.1
Copyright (C) 2022 g10 Code GmbH
License GNU GPL-3.0-or-later <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Home: /home/user/.gnupg
Supported algorithms:
Pubkey: RSA, ELG, DSA, ECDH, ECDSA, EDDSA
Cipher: IDEA, 3DES, CAST5, BLOWFISH, AES, AES192, AES256, TWOFISH,
        CAMELLIA128, CAMELLIA192, CAMELLIA256
Hash: SHA1, RIPEMD160, SHA256, SHA384, SHA512, SHA224
Compression: Uncompressed, ZIP, ZLIB, BZIP2
```

Output is:

```
+ '[' 2 -ne 2 ']'
+ KEY_NAME='Your Name'
+ KEY_EMAIL=your.name@3mdeb.com
+ KEY_COMMENT='Employee Cert Key'
+ KEY_FILE=your.name@3mdeb.com.asc
++ cat
+ BATCH_FILE_CONTENT='%echo Generating a primary certification key and subkeys for signing and encryption
Key-Type: RSA
Key-Length: 4096
Key-Usage: cert
Expire-Date: 0
Subkey-Type: RSA
Subkey-Length: 4096
Subkey-Usage: sign
Name-Real: Your Name
Name-Comment: Employee Cert Key
Name-Email: your.name@3mdeb.com
%no-protection
%commit
%echo Key generation complete'
+ echo '%echo Generating a primary certification key and subkeys for signing and encryption
Key-Type: RSA
Key-Length: 4096
Key-Usage: cert
Expire-Date: 0
Subkey-Type: RSA
Subkey-Length: 4096
Subkey-Usage: sign
Name-Real: Your Name
Name-Comment: Employee Cert Key
Name-Email: your.name@3mdeb.com
%no-protection
%commit
%echo Key generation complete'
++ gpg --batch --generate-key temp_batch_file.txt
++ grep 'gpg: key'
++ awk '{print $3}'
+ KEY_ID=generation
+ '[' -z generation ']'
+ echo 'Modifying key: generation'
Modifying key: generation
+ gpg --command-fd 0 --edit-key generation
gpg (GnuPG) 2.2.40; Copyright (C) 2022 g10 Code GmbH
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

gpg: key "generation" not found: No public key
+ '[' 2 -ne 0 ']'
+ echo 'Error: Cannot change expiration date for signing subkey.'
Error: Cannot change expiration date for signing subkey.
+ exit 1
```

Let me tinker with the script more, it looks like it is not robust enough.

Reviewer: Ok, please ping me if retest is needed @pietrushnic

Author: @macpijan please check bbc5c62d3b0f (sample debug run in commit message).

Author: @macpijan can you try it again?

Author: @macpijan ping?

Reviewer: It works for me now.

```

Please note your key id printed at the end:

```
(...)
[ultimate] (1). Your Name (Employee Cert Key) <your.name@3mdeb.com>

Your KEY_ID: D9E4EB63705C3897
```

### Encrypt revocation cert

Import 3mdeb Founder key:

```shell

> gpg --import key/owner-key/piotr-krol-key.asc

Reviewer: ```suggestion
gpg --import keys/owner-key/piotr-krol-key.asc
```

> gpg --import key/owner-key/piotr-krol-key.asc

Author: @krystian-hebel I would expect 3mdeb-secpack to have a higher priority and trustworthiness than the public server. In the very long run, especially if 3mdeb will grow, I would expect 3mdeb to have its own keyserver as a source of truth.

I'm unsure how `gpg --search-keys` better solves the expired keys problem. Can you elaborate on that argument?

In the case of non-published keys, I know you want to introduce publication as a policy, but we have no such policy right now.  @macpijan @miczyg1 @TomaszAIR If we agree that keys publication of 3mdeb employees is mandatory, then maybe let's add that to scripts. If we all agree, the next question would be which server we should choose. I guess in all those, we can follow Qubes OS's best practices. The key question is, are all Qubes OS developer's keys are published and to which server?

> gpg --import key/owner-key/piotr-krol-key.asc

Reviewer 3: `gpg --search-keys` shows all keys for given search criteria (e.g. email) together with information such as expiration date. That being said, I don't think I ever used `gpg --import` to import expired keys, maybe it also shows warnings or errors.

As for publication of keys, I asked for it because it would help to get better CI for building Qubes OS packages. Right now I had to disable signature checks altogether (https://github.com/TrenchBoot/.github/pull/4/files#diff-42a392ac75e0d1fba3d4468c227c6f1e8776a783c23b105387a12cb0351d298dR78) because there is no way of getting GPG public key out of GitHub.

I agree that 3mdeb-secpack should have a higher priority. As for trustworthiness, it doesn't necessarily depend on us, it depends on policy of whoever wants to use those keys. I can imagine some companies would prefer existing public servers.

> If we agree that keys publication of 3mdeb employees is mandatory, then maybe let's add that to scripts.

How about making it part of CI executed when signed keys are uploaded? It should be possible with https://keys.openpgp.org/about/api, namely `POST /vks/v1/upload` and `POST /vks/v1/request-verify`. Key owners would have to just click the link in the mail they receive to verify.

> gpg --import key/owner-key/piotr-krol-key.asc

Author: > gpg --search-keys shows all keys for given search criteria (e.g. email) together with information such as expiration date. That being said, I don't think I ever used gpg --import to import expired keys, maybe it also shows warnings or errors.

No import does not show that:

```shell
[22:14:06] user:3mdeb-secpack git:(add_proce_for_revocation) $ gpg --import keys/employees-keys/mike-banon-key-signed.asc 
gpg: key D9CE30E2848FE512: 1 signature not checked due to a missing key
gpg: key D9CE30E2848FE512: public key "Mike Banon <mike.banon@3mdeb.com>" imported
gpg: Total number processed: 1
gpg:               imported: 1
gpg: no ultimately trusted keys found
```

```shell
[22:15:13] user:3mdeb-secpack git:(add_proce_for_revocation) $ gpg --list-key mike.banon             
pub   rsa4096 2021-05-11 [SC] [expired: 2023-05-11]
      2CD82FED62A81BA21F31D675D9CE30E2848FE512
uid           [ expired] Mike Banon <mike.banon@3mdeb.com>
```

I guess that replacement makes sense.  I will do that for my key, but I know that not all 3mdeb Team Leaders and Managers have keys on the server. Fixed in 4a060c9c243a47cb3173435409d3c0444f7ac205

The problem is that 3mdeb employees really have to upload keys.

> How about making it part of CI executed when signed keys are uploaded? It should be possible with https://keys.openpgp.org/about/api, namely POST /vks/v1/upload and POST /vks/v1/request-verify. Key owners would have to just click the link in the mail they receive to verify.

This is a very good idea, and I agree with implementing it, but it is out of scope for this PR.  I even proposed it in April #75

> gpg --import key/owner-key/piotr-krol-key.asc

Author: @krystian-hebel can we close this conversation, or you would expect additional improvements?

```

Import your 3mdeb Team Leader or Manager key:

```shell

> gpg --import key/employees-keys/your.tl-or-mgr-name@3mdeb.com

Reviewer: ```suggestion
gpg --import keys/employees-keys/your.tl-or-mgr-name@3mdeb.com
```

```

Encrypt your revocation certificate (please use your KEY_ID from previous
point):

```shell
gpg --encrypt -r piotr.krol@3mdeb.com -r your.tl-or-mgr-name@3mdeb.com -o \
your.name@3mdeb.com.rev.gpg ${HOME}/.gnupg/openpgp-revocs.d/*KEY_ID.rev

```

> Output shouold look as follows:

Reviewer: Instead of pointing out typos, I have enabled pre-commit in this repo: https://github.com/3mdeb/3mdeb-secpack/pull/102/files

Author: @macpijan I added a review to #102

Reviewer: @pietrushnic responded there

Author: @macpijan IIUC we need to merge #102 to close this conversation?

Author: @macpijan ping?

Check signatures after signing:

```shell
gpg --check-sigs your.name@3mdeb.com
```

Output:

> ```shell

Reviewer 3: We're overusing `shell`

```suggestion
```text
```

Make sure your
[Github](https://docs.github.com/en/github/authenticating-to-github/telling-git-about-your-signing-key#telling-git-about-your-gpg-key-2)
and
[Gitlab](https://docs.gitlab.com/ee/user/project/repository/gpg_signed_commits/index.html#adding-a-gpg-key-to-your-account)
is set up correctly to provide GPG commit verification.

### Create a pull request with key

After everything is set up, please issue a pull request to this repo with
signed `your-name-key-signed.asc`.

### Upload to keys.opengpg.org

* Go to https://keys.openpgp.org.
* Upload Your Key:
  * On the main page, you'll see an "Upload" section.
  * Click on the "Choose File" button and select the your.name@3mdeb.com.asc
    file you just created.
  * Click on the "Upload" button.
* Verify Your Email Address:
  * https://keys.openpgp.org will send you an email to verify that the email
    address associated with the key is valid.
  * Open the email and click on the verification link.
  * Once verified, your key will be added to the keys.openpgp.org keyserver.
* After uploading and verifying, you can search for your key on the keyserver
  using your email address to ensure it's been uploaded correctly.

## FAQ

> ### What to do when this key expires?

Reviewer 3: This should also mention that renewed keys have to be sent to GitLab and GitHub. However, as I learned yesterday, you can't just upload new key with the same key ID to GitHub, it will be silently ignored. You have to remove old one and then import the new one. I don't know if GitLab has the same problem.

Author: @krystian-hebel, thanks for the comment. I tried to address it here da0880b8de12d32c49d1f3637f30cad16bafaee1
If you think something should be improved, please provide an actionable description of what can be changed.

Please don't bother about spelling. The automatic spellchecker should fix that. I cannot afford your rate for spellchecking.

In the second case, you should revoke the key. The procedure is
[here](https://superuser.com/a/1526287).

### If we extend expiry every time, why not set "key does not expire"? 

Because people forget about keys or stop using them for many reasons. If
someone extends expiry time, it means someone cares, and the key is still in
use.

### What to do when primary/master private key was lost, compromised or will no longer be used? 

Following guide was made for:
- recovery in case of loosing access to you primary/master private key,
- evidence of primary/master key compromise,

> - event when employee of 3mdeb finished his/her cooperation and should no loger

Reviewer 3: ```suggestion
- event when employee of 3mdeb finished his/her cooperation and should no longer
```

  sign 3mdeb related development work.

Guide is made with assumption, that you still have access to:
- private key or at least previously generated revocation certificate - please

>   note that gpg save revocation certificate automatically on machine you

Reviewer: ```suggestion
  note that gpg saves revocation certificate automatically on machine you
```

  generated key in `/home/USERNAME/.gnupg/openpgp-revocs.d/KEYID.rev`
- list of servers to which your key was uploaded - in theory servers should
  sync between each other,
- list of keys you signed - check below how to obtain it,

Correct key management would require use of subkeys and private part of your

> master key should never be lost because it should stored in safe location.

Reviewer: ```suggestion
master key should never be lost because it should be stored in a safe location.
```

#### Inform your Team Leader or Manager

Please immiedietly let know your 3mdeb Team Leader or Manager about potential
problem with your GPG keys. Delaying this process you put 3mdeb at risk of
compromise.

#### Generate revocation certificate

If you don't have revocation certificate, but have access to private key please
generate revocation certificate by using:

```shell

> gpg --output revoke.asc --gen-revoke KEYID

Reviewer: ```suggestion
gpg --output your.name@3mdeb.com.rev --gen-revoke KEYID
```

Reviewer: It asks me for a reason here:

```
Create a revocation certificate for this key? (y/N) y
Please select the reason for the revocation:
  0 = No reason specified
  1 = Key has been compromised
  2 = Key is superseded
  3 = Key is no longer used
  Q = Cancel
(Probably you want to select 1 here)
Your decision? 
```

This command only generates certificate, not actually revokes the key, right?

I assume we should gather these from current employees?

Author: > This command only generates certificate, not actually revokes the key, right?

@macpijan Yes, it generates a revocation certificate, which can then be applied to the public key. At least, this is how I understand it. We have two situations: either we know what the reason is, or we don't. A further case is our situation when we generate certificates for new employees. If something happens with the key, we know the reason and should use the correct one. We are always able to provide a description after selecting a reason.

> I assume we should gather these from current employees?

Yes.

Please let m know if this section need improvement.

Author: @macpijan ping?

```

Where `KEYID` is the id of the key which you would like to revoke.

#### Revoke your key

```shell
gpg --import revoke.asc
```

#### Upload revoked key to the servers

> Following set of commands assume that machine on which you applied revocation

Reviewer: ```suggestion
Following this set of commands assumes that the machine on which you applied revocation
```

> and machine which you publish revoked key are different. If it is not the case

Reviewer: ```suggestion
and machine on which you publish revoked key are different. If it is not the case,
```

you can skip export step.

Export revoked public key:

```shell
gpg --export -a KEYID > KEYID_revoked.asc
```

Move `KEYID_revoked.asc` to upload machine and import

```shell
gpg --import KEYID_revoked.asc
```

Send to server:

```shell
gpg --keyserver KEYSERVER --send-keys KEYID
```

Where typical `KEYSERVER` would be `https://keyserver.ubuntu.com` and
`https://keys.openpgp.org`.

#### New keys

Before proceeding create new keys following [standard
procedure](#adding-key-to-repository).

#### Revoke key in Thunderbird

* Import revocation certificate into Thunderbrid
  * Tools > OpenPGP Key Manger > File > Import Revocation(s) From File
* If your new key was uploaded to keys.opengpg.org use Keyserver > Discover
  Keys Online
* Replace key in Settings > Account Settings, choose your account End-To-End
  Encryption > Add Key ... > Use your external key through GnuPG, hit Continue
  and provide fingerprint of your certifying key.

#### Gitlab, Github and Gitea revocation

> Go to all wesbites, upload your new public key and delete old one.

Reviewer 3: This would mark all commits as unverified. Maybe we should (if key is not lost) change validity period of the key to last known safe date and resend that to Git.* before creating a new key?

Depending on timing and reach of compromise, there are additional steps, like re-signing potentially compromised commits and tags **after confirming that they are good, i.e. after review**, and ideally creating a public announcement about why this was done, especially if any release is impacted by that.

Author: > This would mark all commits as unverified. Maybe we should (if key is not lost) change validity period of the key to last known safe date and resend that to Git.* before creating a new key?

I will check that and see if shortening the expiration date is possible, but I have some doubts about that. I probably will ask Marek about his opinion. The Qubes OS Team doesn't care much about Github green marks. They check everything themselves. We care about markings, but maybe we shouldn't because that green marking depends on Github proprietary implementation, and Github seems not to care about supporting edge cases well enough to respect the security procedures they present to users. Or maybe there is no business value in correctly handling events like key expiration, revocation by the web of trust, loss, or compromise.

> Depending on timing and reach of compromise, there are additional steps, like re-signing potentially compromised commits and tags after confirming that they are good, i.e. after review, and ideally creating a public announcement about why this was done, especially if any release is impacted by that.

The key question is if we assume compromise without evidence of compromise. This also leads to the question of how much we trust in the protection provided by HSMs. For example, the key was lost on Nitrokey or Yubikey. Are we assuming those are exploitable?

Please let me know what action you expect here because I cannot comprehend your comment. Writing comprehensive security policies for hypothetical events does not have a high enough expected value (cost of writing multiplied by probability of event). What we can do is act on what happened. Extensive holistic design is beyond our abilities, and I'm pretty sure it's not our problem. I would be glad to be wrong here and get examples of how others deal with what you propose.

Reviewer 3: > We care about markings, but maybe we shouldn't because that green marking depends on Github proprietary implementation, and Github seems not to care about supporting edge cases well enough to respect the security procedures they present to users.

In that case there is no reason to upload GPG keys to GH in the first place, right?

> Please let me know what action you expect here because I cannot comprehend your comment.

If you know that someone *could* have gained access to your key (lost HSM, PC left unlocked without your presence, someone asked you to `curl ... | sh` etc.), but you're not sure whether it was on Thursday or Friday, you assume that all commits created and signed on or before Wednesday are valid. Thursday and newer have to be re-signed with new key, because they could have been signed by you or by anyone who obtained the key. It shouldn't matter whether it was on HSM or not, but I guess if someone uses *only* HSM to sign with no backup stored (and used) elsewhere it shouldn't be possible, as long as it is impossible to get the key out of token without proper authentication.

Author: > In that case there is no reason to upload GPG keys to GH in the first place, right?

By checking random repos of Qubes OS, I see everything marked with green, so I assume we should also care. AFAIK someone pointed out some of our commits are not signed, despite those being signed, but GH didn't show a signature for some reason. Let me try to contact @marmarek and ask for his opinion on the expiration trick that should be used.

> If you know that someone could have gained access to your key (lost HSM, PC left unlocked without your presence, someone asked you to curl ... | sh etc.), but you're not sure whether it was on Thursday or Friday, you assume that all commits created and signed on or before Wednesday are valid. Thursday and newer have to be re-signed with new key, because they could have been signed by you or by anyone who obtained the key. It shouldn't matter whether it was on HSM or not, but I guess if someone uses only HSM to sign with no backup stored (and used) elsewhere it shouldn't be possible, as long as it is impossible to get the key out of token without proper authentication.

I get all you wrote and agree, but I still don't understand what action you expect. Do you want to add some statements explaining the above to the documentation? In my case, we know when the key was used physically last time. There are no commits after that time, which I'm aware of. All new commits from me rely on the new key, verified as it would be a new key.

According to the new methodology of key creation, we should use the Certify key only in a safeguarded location. You can use subkeys in various locations, which can be easily revoked without undermining the web of trust built on your key.

Reviewer 4: We "care" about green marks only because it's easy for contributors to see what is going on - how signature should look like and whether they did it correctly. But for example we have our own CI script for checking signatures anyway (if it fails, it shows a link to a documentation how to sign commits). And ofc we don't depend on those green marks for any kind of integrity.

BTW github green mark is also on commits you do on github's web interface, which is very different trust model (_if_ you trust github, it's github certifying by which user the commit was created, in a way it's verifiable outside of github too). Generally we don't accept those, but in practice we accept it for minor changes (fixing typos in comments, documentation etc).

Reviewer 3: > Do you want to add some statements explaining the above to the documentation?

This would be nice.

> According to the new methodology of key creation, we should use the Certify key only in a safeguarded location. You can use subkeys in various locations, which can be easily revoked without undermining the web of trust built on your key.

Can you add a diagram that would represent the new hierarchy of keys?

Author: @marmarek, thanks for the explanation. @krystian-hebel it would be great if you could express what are your expectations to satisfy this conversation.

Author: > Can you add a diagram that would represent the new hierarchy of keys?

Yes, I can, but it is trivial and visible in gpg output. We use here quite common practice among advanced gpg users:

```
pub   rsa4096 2023-10-11 [C] <- Primary/Master Key for Certifying (aka signing others keys)
      5468873B74F163152785D2CC67D4F3E372CBC3A9
uid           [ultimate] Piotr Król (Employee Cert Key) <piotr.krol@3mdeb.com>
sig 3        67D4F3E372CBC3A9 2023-10-11  Piotr Król (Employee Cert Key) <piotr.krol@3mdeb.com>
sig          9963C36AAC3B2B46 2023-10-12  Maciej Pijanowski <maciej.pijanowski@3mdeb.com>
sub   rsa4096 2023-10-11 [S] [expires: 2024-10-10] <- Signing Subkey (key for signing git commits and probably other stuff)
sig          67D4F3E372CBC3A9 2023-10-11  Piotr Król (Employee Cert Key) <piotr.krol@3mdeb.com>
sub   rsa4096 2023-10-11 [E] [expires: 2024-10-10] <- Encryption Subkey (key for email encryption)
sig          67D4F3E372CBC3A9 2023-10-11  Piotr Król (Employee Cert Key) <piotr.krol@3mdeb.com>
```

Author: Github documentation:

> To verify as many of your commits as possible, you can add expired and revoked keys. If the key meets all other verification requirements, commits that were previously signed by any of the corresponding private keys will show as verified and indicate that their signing key is expired or revoked.

https://docs.github.com/en/authentication/managing-commit-signature-verification/adding-a-gpg-key-to-your-github-account

So the question is what should be our policy in the above context?

Reviewer 3: Yes, this is how I had to upload my key after extending its expiry date. Without removing it first, new key was silently ignored, with no warning or error whatsoever. My old commits are still green so I guess it worked.

Author: @marmarek @krystian-hebel ok, I probably made another mistake when I received `We got an error adding your GPG key. Please verify the input is a valid GPG key.`, but now I can upload my revoked key. Now, in light of the above statuses does it make sense to modify the expiration date? No matter if the key is expired or revoked Github shows the signature as verified.

Author: It is possible to have key both revoked and expired:

![gh_rev_exp](https://github.com/3mdeb/3mdeb-secpack/assets/1071894/36850330-db3c-46c3-bf05-c102af07aeb4)

Reviewer 3: `The key expired after the commit was signed` - does it still show as verified if commit is created **after** the expiration date?

Author: @krystian-hebel improved description that covers expiration date calculation and set up as well as upload to Github: fbb2fa6f04b2

Please let me know if that is enough or if you expect more changes.

Reviewer 3: I've added comment in https://github.com/3mdeb/3mdeb-secpack/commit/fbb2fa6f04b2649eee91c6f2d383c5d396cb087c, but it isn't visible here. Is that how it's supposed to work? :thinking:

Author: > The key expired after the commit was signed - does it still show as verified if commit is created after the expiration date?

How do you force gpg to use the key that expired for signing?

```
vault% GIT_TRACE=1 git commit -SFD8B7B3145214C77 -m "README.md: sign after key expired 2"
23:32:28.693489 git.c:460               trace: built-in: git commit -SFD8B7B3145214C77 -m 'README.md: sign after key expired 2'
23:32:28.693972 run-command.c:655       trace: run_command: gpg --status-fd=2 -bsau FD8B7B3145214C77
error: gpg failed to sign the data
fatal: failed to write commit object
vault% gpg --status-fd=2 -bsau FD8B7B3145214C77
[GNUPG:] KEYEXPIRED 1697320809
[GNUPG:] KEY_CONSIDERED 04C97C54E51FEBA30806D8FC40BABCE7542B0B4A 3
gpg: skipped "FD8B7B3145214C77": Unusable secret key
[GNUPG:] INV_SGNR 9 FD8B7B3145214C77
[GNUPG:] FAILURE sign 54
gpg: signing failed: Unusable secret key
```

Author: > I've added comment in https://github.com/3mdeb/3mdeb-secpack/commit/fbb2fa6f04b2649eee91c6f2d383c5d396cb087c, but it isn't visible here. Is that how it's supposed to work? 🤔

@krystian-hebel, I would say no if we consider how GH works right now. A quick look at the commit may work if you want to solve a conversation. If you want to start a new conversation, you should do that in review; otherwise, we lose track. I guess we already had that discussion in 3mdeb internally and decided we favor Gerrit's behavior over Github precisely because of that.

Author: @krystian-hebel I addressed issues in fbb2fa6f04b2649eee91c6f2d383c5d396cb087c commit

Reviewer 3: > How do you force gpg to use the key that expired for signing?

Not tested, but maybe https://superuser.com/a/1505607 would work, just set `gpg.program=gpg --faked-system-time 20231005T000000`. Not sure if this would change commit date as well, but I don't think so. If not, you can sign the commit while the key is still valid and change expiration date to the past after that. There is a lot of time traveling involved :upside_down_face:

Maybe `--faked-system-time` would make expiring keys in the past easier?

Author: OMG, we are really beating this horse to death. I really wonder about ROI, but it is better to be safe than sorry.

It was not as easy as you described because gpg.program doesn't like params, but put stuff in the wrapper script, and it worked.

```shell
vault% gpg --faked-system-time 20231005T000000 --list-keys 04C97C54E51FEBA30806D8FC40BABCE7542B0B4A
gpg: WARNING: running with faked system time: 2023-10-05 00:00:00
pub   rsa4096 2022-10-14 [C]
      04C97C54E51FEBA30806D8FC40BABCE7542B0B4A
uid           [ultimate] Piotr Król Signature After Expiration Date (Employee Cert Key) <piotr.krol@3mdeb.com>
sub   rsa4096 2022-10-14 [S] [expires: 2023-10-14]
sub   rsa4096 2022-10-14 [E] [expires: 2023-10-14]

vault% GIT_TRACE=1 git commit -S04C97C54E51FEBA30806D8FC40BABCE7542B0B4A -m "README.md: sign with expired key after expiration"
23:27:25.574281 git.c:460               trace: built-in: git commit -S04C97C54E51FEBA30806D8FC40BABCE7542B0B4A -m 'README.md: sign with expired key after expiration'
23:27:25.574781 run-command.c:655       trace: run_command: /home/user/gpg.sh --status-fd=2 -bsau 04C97C54E51FEBA30806D8FC40BABCE7542B0B4A
23:27:25.610630 run-command.c:655       trace: run_command: git maintenance run --auto --no-quiet
23:27:25.620425 git.c:460               trace: built-in: git maintenance run --auto --no-quiet
[main 32a404f] README.md: sign with expired key after expiration
 1 file changed, 2 insertions(+)
vault% qvm-copy ../gh-gpg-testing-repo 
sent 33/194 KB
vault% gpg --list-keys 04C97C54E51FEBA30806D8FC40BABCE7542B0B4A                            
pub   rsa4096 2022-10-14 [C]
      04C97C54E51FEBA30806D8FC40BABCE7542B0B4A
uid           [ultimate] Piotr Król Signature After Expiration Date (Employee Cert Key) <piotr.krol@3mdeb.com>
```

But the effect is not the one we would expect because it claims the same thing: `This commit was signed with the committer’s verified signature. The key expired after the commit was signed.`.  This makes sense because despite the commit was made on 18th Oct: https://github.com/pietrushnic/gh-gpg-testing-repo/commit/32a404f8c04442965eafc4b71d240c25a1d3fd60 signature of this commit was made 5th October. So, definitely time travelers.

```
--expert
--no-expert

    Allow the user to do certain nonsensical or "silly" things like signing an expired or revoked key, or certain potentially incompatible things like generating unusual key types. This also disables certain warning messages about potentially incompatible actions. As the name implies, this option is for experts only. If you don’t fully understand the implications of what it allows you to do, leave this off. --no-expert disables this option.
```

But id doesn't work as explained.

```
vault% gpg --expert --status-fd=2 -bsau 04C97C54E51FEBA30806D8FC40BABCE7542B0B4A 
[GNUPG:] KEYEXPIRED 1697320809
[GNUPG:] KEY_CONSIDERED 04C97C54E51FEBA30806D8FC40BABCE7542B0B4A 3
gpg: skipped "04C97C54E51FEBA30806D8FC40BABCE7542B0B4A": Unusable secret key
[GNUPG:] INV_SGNR 9 04C97C54E51FEBA30806D8FC40BABCE7542B0B4A
[GNUPG:] FAILURE sign 54
gpg: signing failed: Unusable secret key
```

Five months ago, there was this option added `--debug-ignore-expiration`: https://github.com/gpg/gnupg/commit/c30d5829c9ab076e9fd39c223273b065cdb58a0d
It requires gpg 2.4.2

That option allowed me to bypass checks and signs, but GitHub detects it like above. So maybe that option behind the scene use faketime, which does not help in performing that test.

@krystian-hebel, how would you like to proceed from this point?

Author: @krystian-hebel ?

Reviewer 3: > But the effect is not the one we would expect because it claims the same thing: This commit was signed with the committer’s verified signature. The key expired after the commit was signed.. This makes sense because despite the commit was made on 18th Oct: https://github.com/pietrushnic/gh-gpg-testing-repo/commit/32a404f8c04442965eafc4b71d240c25a1d3fd60 signature of this commit was made 5th October. So, definitely time travelers.

This is FUBAR. It means that when your key gets stolen, even after its expiry date someone can use it to sign a commit that will be green in GUI. Even in `git log --show-signature` you have to pay attention to the dates, because this isn't immediately visible. Do you feel like reporting it to GH/git?

This has gone slightly off-topic. I'm OK with leaving the content in the current state, so feel free to mark this as resolved, unless you want to add anything.

Author: @krystian-hebel, if I knew you wanted to go down the rabbit hole, I would insist on moving that to issue.

>  This is FUBAR. It means that when your key gets stolen, even after its expiry date someone can use it to sign a commit that will be green in GUI. Even in git log --show-signature you have to pay attention to the dates, because 
this isn't immediately visible. 

I will leave some logs:

```shell
commit 6ef3aa2e8186bafa168a1451e78346d018b20c4e (HEAD -> main, origin/main)
gpg: Signature made Thu 19 Oct 2023 01:07:59 AM CEST
gpg:                using RSA key 2BF8D9D82D872920449A79C9FD8B7B3145214C77
gpg: Good signature from "Piotr Król Signature After Expiration Date (Employee Cert Key) <piotr.krol@3mdeb.com>" [unknown]
gpg: Note: This key has expired!
Primary key fingerprint: 04C9 7C54 E51F EBA3 0806  D8FC 40BA BCE7 542B 0B4A
     Subkey fingerprint: 2BF8 D9D8 2D87 2920 449A  79C9 FD8B 7B31 4521 4C77
Author: Piotr Król <piotr.krol@3mdeb.com>
Date:   Thu Oct 19 01:07:59 2023 +0200

    README: sign after key expiration 2
```

```shell
[22:43:24] user:gh-gpg-testing-repo git:(main) $ gpg --edit-key 04C97C54E51FEBA30806D8FC40BABCE7542B0B4A 
gpg (GnuPG) 2.2.40; Copyright (C) 2022 g10 Code GmbH
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Secret subkeys are available.

pub  rsa4096/40BABCE7542B0B4A
     created: 2022-10-14  expires: never       usage: C   
     trust: unknown       validity: unknown
sub  rsa4096/FD8B7B3145214C77
     created: 2022-10-14  expired: 2023-10-14  usage: S   
sub  rsa4096/C579B798739ACC35
     created: 2022-10-14  expired: 2023-10-14  usage: E   
[ unknown] (1). Piotr Król Signature After Expiration Date (Employee Cert Key) <piotr.krol@3mdeb.com>

gpg> 
```

@marmarek, what is your opinion about that weird behavior?

Author: @krystian-hebel I guess there is no way to submit issues, just discussion: https://github.com/orgs/community/discussions/72294

Author: Report sent also using: https://support.github.com/contact/bug-report

Reviewer 4: In principle you can't really know when the signature was made without some external authority. In this case you could use a hint on commit date, but this is easy to fake too.

Author: @marmarek The key question is whether GH behavior is really FUBAR, as @krystian-hebel said. Dates can be faked, but that does not explain why showing a green badge when signature content clearly says it is invalid.

I wonder if I can force signing with a revoked key and what GH will do in that situation. Does the signature made with the revoked key showing verified big enough SNAFU?

Reviewer 3: > I wonder if I can force signing with a revoked key and what GH will do in that situation.

This may be possible if key is revoked on one GPG instance and sent to GH, and another GPG is used to sign new commits.

Author: The key question is what will be shown by GH. I received an answer from GH's first-line

![gpg_verified_badge_gh_1st_line_resp](https://github.com/3mdeb/3mdeb-secpack/assets/1071894/3e826b09-9b7f-44a6-85cd-675c5e7bfda4)
 support:

Author: It doesn't look like it will be escalated. This thread probably deserves a blog post. I will add it to my backlog.

Reviewer 5: Hey can I somewhere read more about this topic? (GH GPG)

Author: @bukowa, what more would you like to read? Everything is in this thread. There is nothing more.

The repository where I did experiments is https://github.com/pietrushnic/gh-gpg-testing-repo - but I removed experimental keys from my configuration. I'm pretty sure that reading the above thread carefully, you can repeat all I reported.

#### Obtain list of keys you signed

Following procedure looks through keyring, so it should be prepared in
evironment which is used to signing operations. For Qubes OS it would be vault
VM.

```shell
for i in $(gpg --list-keys --with-colons | awk -F: '/^fpr/ {print $10}'|xargs);do
	if gpg --list-sigs $i|grep -q -E "^sig\s+$(echo KEYID | tail -c 17)";then
		echo $i
	fi
done
```

> List of keys have to be signed by new key.

Reviewer: ```suggestion
It produces a list of keys that have to be signed by a new key.
```
